### PR TITLE
Add voting power based fund distribution contract.

### PR DESCRIPTION
**Pull request by ezekiiel**
This PR adds a fund distribution contract wherein funds are distributed proportional to the amount of voting power a member has at a given block height. This may be used to liquidate and shut down a DAO or to pay out dividends to DAO members.

The contract can distribute an arbitrary number of native or cw20 tokens to an arbitrary number of DAO members. Its distribution methods attempt to distribute all tokens by default, but have pagination in case the number of tokens being distributed hits gas limits (read: many cw20 send messages due to large DAO treasury).

An optional admin can be set for this contract. That admin may withdraw tokens that have been allocated to this distribution contract. Any address may provide funds for distribution by this contract though it is imagined that it will be the DAO doing this.

At the moment it is possible to fund the contract more than once with the same token. For example:

1. DAO with two equally weighted members pays out 100 Juno to the contract.
2. Member 1 claims and receives 50 Juno.
3. DAO pays out an additional 100 Juno.
4. Member 1 claims and receives 50 more Juno.
5. Member 2 claims and receives 100 Juno.

Another possible ordering that addresses an edge case:

1. DAO with two equally weighted members pays out 100 Juno to the contract.
2. Member 1 claims and receives 50 Juno.
3. DAO withdraws the initial Juno allocation.
4. DAO provides 200 Juno.
5. Member 1 claims and receives 50 more Juno.
6. Member 2 claims and receives 100 Juno.

*TODO:* unaddressed edge case:

1. DAO with two equally weighted members pays out 100 Juno to the contract.
2. Member 1 claims and receives 50 Juno.
3. DAO withdraws the initial Juno allocation.
4. DAO provides 50 Juno.
5. Member 1 claims and receives 25 more Juno.
6. Member 2 claims and receives 25 Juno.

### CODE REVIEW

File: contracts/cw-fund-distributor/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,621 @@
...
+                .unwrap_or_default();
+            let entitled = total_entitlement - claimed;
+            Ok((key, entitled))
+        })
+        .collect::<StdResult<Vec<_>>>()
+}
+
+/// Generic function for paginating a list of (K, V) pairs in a
+/// CosmWasm Map.
+fn paginate_items<'a, K, V>(
```
Comment (Callum-A) [inline contracts/cw-fund-distributor/src/contract.rs:153]: i can totally see us wanting to move this to a shared utils at a later date

**Comment by JakeHartnell** [comment]
Assume this would close #266?

File: contracts/cw-fund-distributor/src/contract.rs (hunk 1 above)
Comment (ezekiiel) [reply contracts/cw-fund-distributor/src/contract.rs:153]: yes!

**Comment by ezekiiel** [comment]
> @JakeHartnell  Assume this would close https://github.com/DA0-DA0/dao-contracts/issues/266?

yep!

**Comment by ezekiiel** [comment]
Closing this for now as immediate focus is elsewhere and don't want to clutter the PR queue.

**Pull request closed without merging**